src/main/java/org/wildfly/security/auth/provider/JAASSecurityRealm.java

```diff
@@ -0,0 +1,244 @@
+                    // preserve backwards compatibility: custom handlers were allowed in the past as long as they had a public setSecurityInfo method.
+                    final Method setSecurityInfo = handler.getClass().getMethod("setSecurityInfo", Principal.class, Object.class);
+                    setSecurityInfo.invoke(callbackHandler, this.principal, credential);
+                    return callbackHandler;
+                } catch (Exception e) {
+                    throw ElytronMessages.log.failedToInstantiateCustomHandler(e);
+                }
+            }
+        }
+    }
+
+    private class DefaultCallbackHandler implements CallbackHandler {
+
+        private final Principal principal;
+        private final Object credential;
+
+        private DefaultCallbackHandler(final Principal principal, final Object credential) {
+            this.principal = principal;
+            this.credential = credential;
+        }
+
+        @Override
+        public void handle(Callback[] callbacks) throws IOException, UnsupportedCallbackException {
+            if (callbacks == null)
+                throw ElytronMessages.log.invalidNullCallbackArray();
+
+            for (Callback callback : callbacks) {
+                if (callback instanceof NameCallback) {
+                    NameCallback nameCallback = (NameCallback) callback;
+                    if (principal != null)
+                        nameCallback.setName(this.principal.getName());
+                }
+                else if (callback instanceof PasswordCallback) {
+                    PasswordCallback passwordCallback = (PasswordCallback) callback;
+                    if (this.credential instanceof char[]) {
+                        passwordCallback.setPassword((char[]) credential);
+                    }
+                    else if (this.credential instanceof String) {
+                        passwordCallback.setPassword(((String) credential).toCharArray());
+                    }
+                    else if (this.credential instanceof ClearPassword) {
+                        passwordCallback.setPassword(((ClearPassword) credential).getPassword());
+                    }
+                    else {
+                        throw ElytronMessages.log.failedToConvertCredentialToPassword(callback);
+                    }
+                }
+                else {
+                    throw ElytronMessages.log.unsupportedCallback(callback);
+                }
+            }
+        }
+    }
+
+    private class JAASAuthenticatedRealmIdentity implements AuthenticatedRealmIdentity {
+
+        private final Principal principal;
+
+        private final Subject subject;
+
+        private JAASAuthenticatedRealmIdentity(final Principal principal, final Subject subject) {
+            this.principal = principal;
+            this.subject = subject;
+            // todo investigate subject for a caller principal for backwards compatibility.
```

Comment:
PicketBox modules have used a special group called CallerPrincipal that allows a module to specify the identity that should be used by the container in methods like getCallerPrincipal(). I think the AuthenticatedRealmIdentity should be using that principal if available.